README.md
<!-- 
BSD 3-Clause License
Copyright (c) 2022, GM Consult Pty Ltd
All rights reserved. 
-->

[![GM Consult Pty Ltd](https://raw.githubusercontent.com/GM-Consult-Pty-Ltd/gmconsult_dev/main/doc/images/gmconsult_dev_header.png?raw=true "GM Consult Pty Ltd")](https://github.com/GM-Consult-Pty-Ltd)
## **Lubricate / standardize unit testing.**

*THIS PACKAGE IS A PROPRIETARY, UNLISTED RESOURCE AND SUBJECT TO BREAKING CHANGES WITHOUT NOTICE.*

Skip to section:
- [Overview](#overview)
- [Usage](#usage)
- [API](#api)
- [Definitions](#definitions)
- [References](#references)
- [Issues](#issues)

## Overview

A collection of utilities used for unit testing and code generation.  May require knowledge of
GM Consult coding practices and management systems:
* Use the [Console](https://pub.dev/documentation/gmconsult_dev/latest/gmconsult_dev/Console-class.html) class to print a collection JSON documents to the console as a formatted table.
* Use the [SaveAs](https://pub.dev/documentation/gmconsult_dev/latest/gmconsult_dev/SaveAs-class.html) class to save text and JSON documents to disk.
* Use datasets from the [TestData](https://pub.dev/documentation/gmconsult_dev/latest/gmconsult_dev/TestData-class.html) class in your unit tests.
* Use the [JsonDataService](https://pub.dev/documentation/gmconsult_dev/latest/gmconsult_dev/JsonDataService-class.html) class to quickly create and populate a asynchronous persisted datastore.
* Use the static methods from [API](https://pub.dev/documentation/gmconsult_dev/latest/gmconsult_dev/SaveAs-class.html) to `GET` or `POST` from/to a RESTful API.

## Usage

Fully worked [examples](https://pub.dev/packages/gmconsult_dev/example) are included in the package repository.

In the `pubspec.yaml` of your flutter project, add the following dev dependency:

```yaml
dev_dependencies:
  gmconsult_dev: <latest_version>
```

In your test file add the following import:

```dart
// import the `Console`, `SaveAs` and `JsonDataService` classes
import 'package:gmconsult_dev/gmconsult_dev.dart';

// import the typedefs if needed
import 'package:gmconsult_dev/type_definitions.dart';

// import the `TestData` datasets if needed
import 'package:gmconsult_dev/test_data.dart';
```

### Print to the console with [Console](https://pub.dev/documentation/gmconsult_dev/latest/gmconsult_dev/Console-class.html)

Pass a collection of JSON documents to the [Console.out()](https://pub.dev/documentation/gmconsult_dev/latest/gmconsult_dev/Console/out.html) static method:
```dart

// a sample list of JSON documents
  const jsonCollection = {
    'term0': {'term': 'bodrer', 'dL': 0.0, 'cLs': 1.0},
    'term1': {'term': 'board', 'dL': 0.26303837, 'cLs': 0.736},
    'term2': {'term': 'border', 'dL': 0.0, 'cLs': 1.0},
    'term3': {'term': 'boarder', 'dL': 0.222395656, 'cLs': 0.77761454},
    'term4': {'term': 'brother', 'dL': 0.22239353904, 'cLs': 0.7776134576},
    'term5': {'term': 'broad', 'dL': 0.263035476, 'cLs': 0.73697546},
    'term6': {'term': 'bored', 'dL': 0.2630354, 'cLs': 0.7369766},
  };

  // print a seperator 70 charcters wide
  Console.seperator(70);
  // prints
  // ======================================================================

  // pass a title and results to the Console.out static method to print results
  Console.out(title: 'MY TEST NAME', results: jsonCollection.values.toList());
  // prints:
  //
  // MY TEST NAME
  // ╔═══════════╤═════════════════╤════════════════╗
  // ║  term     │       dL        │       cLs      ║
  // ╟───────────┼─────────────────┼────────────────╢
  // ║  bodrer   │        0.00000  │       1.00000  ║
  // ║  board    │        0.26304  │       0.73600  ║
  // ║  border   │        0.00000  │       1.00000  ║
  // ║  boarder  │        0.22240  │       0.77761  ║
  // ║  brother  │        0.22239  │       0.77761  ║
  // ║  broad    │        0.26304  │       0.73698  ║
  // ║  bored    │        0.26304  │       0.73698  ║
  // ╚═══════════╧═════════════════╧════════════════╝

//
```

### Mock a data service with [JsonDataService](https://pub.dev/documentation/gmconsult_dev/latest/gmconsult_dev/JsonDataService-class.html)

Initialize a [JsonDataService](https://pub.dev/documentation/gmconsult_dev/latest/gmconsult_dev/JsonDataService-class.html), populate it and then read from it.

```dart

  // get some data
  final jsonCollection = {
    'term0': {'term': 'bodrer', 'dL': 0.0, 'cLs': 1.0},
    'term1': {'term': 'board', 'dL': 0.26303837, 'cLs': 0.736},
    'term2': {'term': 'border', 'dL': 0.0, 'cLs': 1.0},
    'term3': {'term': 'boarder', 'dL': 0.222395656, 'cLs': 0.77761454},
    'term4': {'term': 'brother', 'dL': 0.22239353904, 'cLs': 0.7776134576},
    'term5': {'term': 'broad', 'dL': 0.263035476, 'cLs': 0.73697546},
    'term6': {'term': 'bored', 'dL': 0.2630354, 'cLs': 0.7369766},
  };

  // instantiate a JsonDataService
  final service = await JsonDataService.hydrate(
      '${Directory.current.path}\\example', 'sampleStocks');

  // clear the datastore
  await service.dataStore.clear();

  // add all the elements of sampleStocks
  await service.batchUpsert(jsonCollection);

  // read a few records
  final results = (await service.batchRead(['term3', 'term5'])).values.toList();

  // print the records to the console
  Console.out(title: 'JSON DATA SERVICE', results: results);

  // close the service, releasing the resources
  await service.close();

  // prints:
  // ———————————————————————————————————————————————————————————————————————————————————
  // JSON DATA SERVICE                                                                  
  // ___________________________________________________________________________________
  // │  term                  │             dL             │            cLs            │
  // ———————————————————————————————————————————————————————————————————————————————————
  // │  boarder               │                   0.22240  │                  0.77761  │
  // │  broad                 │                   0.26304  │                  0.73698  │
  // ———————————————————————————————————————————————————————————————————————————————————

```

### Save test results [SaveAs](https://pub.dev/documentation/gmconsult_dev/latest/gmconsult_dev/SaveAs-class.html) using [TestData](https://pub.dev/documentation/gmconsult_dev/latest/gmconsult_dev/TestData-class.html)

To save text, JSON or test results, just call the appropriate static method from the [SaveAs](https://pub.dev/documentation/gmconsult_dev/latest/gmconsult_dev/SaveAs-class.html) class. We are using datasets from the [TestData](https://pub.dev/documentation/gmconsult_dev/latest/gmconsult_dev/TestData-class.html) class in the examples below.

```dart

      // save text from a `TestData` dataset
      await SaveAs.text(
        fileName: 'test/data/google',
        text: TestData.text,
      );

      // save a JSON document from a `TestData` dataset
      await SaveAs.json(
        fileName: 'test/data/google',
        json: TestData.json,
      );

      // save your TestResults collection from a `TestData` dataset
      await SaveAs.results(
        fileName: 'test/data/results',
        results: TestData.stockData.values,
        keyBuilder: (json) => json['id'],
      );

```

Use the static methods from [API](https://pub.dev/documentation/gmconsult_dev/latest/gmconsult_dev/API-class.html) to `GET` some data from a RESTful API:

```dart
  // Get some JSON data from a public API endpoint
  final json = await API.get(
      host: 'boredapi.com',
      path: 'api/activity',
      queryParameters: {'key': '5881028'},
      // headers: headers,
      isHttps: true);

  // print the results
  print('Request URL:        ${json['_%url']}');
  print('Response code:      ${json['_%status']}');
  print('Activity:           ${json['activity']}');

  // prints to the console:
  //    Request URL:        https://boredapi.com/api/activity?key=5881028
  //    Response code:      200
  //    Activity:           Learn a new programming language
```

## API

API documentation is available on [pub.dev](https://pub.dev/documentation/gmconsult_dev/latest/).

## Definitions

None

## References

None

## Issues

GM Consult team members should [log issues and bugs](https://github.com/GM-Consult-Pty-Ltd/gmconsult_dev/issues).  

This project is an internal package for GM Consult projects and we cannot respond to pull requests or issues raised by non team members.




CHANGELOG.md
<!-- 
BSD 3-Clause License
Copyright (c) 2022, GM Consult Pty Ltd
All rights reserved. 
-->

*THIS PACKAGE IS A PROPRIETARY, UNLISTED RESOURCE AND SUBJECT TO BREAKING CHANGES WITHOUT NOTICE.*


### 0.0.7+5

#### *Updated*
* Dependencies.

### 0.0.7+4

#### *Updated*
* Documentation.

### 0.0.7+3

#### *Updated*
* License
* Dependencies.
* Tests.
* Examples.
* Documentation.

### 0.0.7+2

#### *Bug fixes*
* Fixed missing white-space in JSON data.

### 0.0.7+1

#### *Bug fixes*
* Fixed alignment of values.

### 0.0.7

#### *Breaking changes*
* Renamed `Echo` class to `Console`.
* Removed the `Console.printResults` method. Use the `Console.out` static method instead.
* Removed the `Console` unnamed factory constructor.

#### *New*
* Added `TestData.kGramIndex` dataset.
* Added `Console.out` static method.
* Added `Console.seperator` static method.

#### *Bug fixes*
* Fixed exception thrown when `API.get` receives a status code other than 200.

#### *Updated*
* Changed formatting of tables in `Console.out` method.
* Dependencies.
* Tests.
* Examples.
* Documentation.


### 0.0.6

#### *New*
* Added `Echo.separator` method.

#### *Updated*
* Dependencies.
* Tests.
* Documentation.

### 0.0.5

#### *New*
* Added `API` class.

#### *Updated*
* Dependencies.
* Tests.
* Documentation.

### 0.0.4

Stable release.

### 0.0.4-4

#### *Updated*
* Dependencies.
* Tests.
* Documentation.

### 0.0.4-3

#### *Updated*
* Dependencies.
* Tests.
* Documentation.

### 0.0.4-2

#### *Updated*
* Dependencies.
* Tests.
* Documentation.

### 0.0.4-1

#### *Updated*
* Dependencies.
* Tests.
* Documentation.

### 0.0.3+1

#### *Updated*
* Dependencies.
* Tests.
* Examples.
* Documentation.

### 0.0.3

#### *New*
* Added `SaveAs` class.
* Added `TestData` class.

#### *Updated*
* Dependencies.
* Tests.
* Examples.
* Documentation.

### 0.0.3-2

#### *New*
* Added `SaveAs` class.
* Added static method `SaveAs.text`.

#### *Updated*
* Dependencies.
* Tests.
* Examples.
* Documentation.

### 0.0.3-1

#### *New*
* Added `SaveAs` class.
* Added static method `SaveAs.text`.

#### *Updated*
* Dependencies.
* Tests.
* Examples.
* Documentation.

### 0.0.2

#### *Breaking changes*
* Minor change to table formatting in `Echo` class

#### *Updated*
* Dependencies.
* Tests.
* Examples.
* Documentation.

### 0.0.1
**BETA 1**

#### *New*
* class `Echo` coded and tested.
* class `JsonDataService` coded and tested.
* Dependencies.
* Tests.
* Examples.
* Documentation.

### 0.0.1-1

* **PRE-RELEASE**
* Initial version.

example/gmconsult_dev_example.dart
// Copyright ©2022, GM Consult (Pty) Ltd.
// BSD 3-Clause License
// All rights reserved

// ignore_for_file: unused_import

import 'dart:io';

// import the `Console`, `SaveAs` and `JsonDataService` classes
import 'package:gmconsult_dev/gmconsult_dev.dart';

// import the typedefs if needed
import 'package:gmconsult_dev/type_definitions.dart';

// import the `TestData` datasets if needed
import 'package:gmconsult_dev/test_data.dart';

void main() async {
  //

  // get some JSON from an API
  await _getHttp();

  const jsonCollection = {
    'term0': {'term': 'bodrer', 'dL': 0.0, 'cLs': 1.0},
    'term1': {'term': 'board', 'dL': 0.26303837, 'cLs': 0.736},
    'term2': {'term': 'border', 'dL': 0.0, 'cLs': 1.0},
    'term3': {'term': 'boarder', 'dL': 0.222395656, 'cLs': 0.77761454},
    'term4': {'term': 'brother', 'dL': 0.22239353904, 'cLs': 0.7776134576},
    'term5': {'term': 'broad', 'dL': 0.263035476, 'cLs': 0.73697546},
    'term6': {'term': 'bored', 'dL': 0.2630354, 'cLs': 0.7369766},
  };

  // print the jsonCollection with the `Console.out` static method
  _printJson(jsonCollection);

  // save data to files using static methods of the `SaveAs` class
  await _saveAsExamples();

  // create a [JsonDataService] using a Hive Box and populate it with
  // jsonCollection
  await _jsonDataServiceExample(jsonCollection);
}

/// Save data to files using static methods of the `SaveAs` class.
Future<void> _saveAsExamples() async {
  // save text
  await SaveAs.text(
    fileName: 'example/data/google',
    text: TestData.text,
  );

  // save a JSON document
  await SaveAs.json(
    fileName: 'example/data/google',
    json: TestData.json,
  );

  // save your TestResults collection
  await SaveAs.results(
    fileName: 'example/data/results',
    results: TestData.stockData.values,
    keyBuilder: (json) => json['id'],
  );
}

/// Get some JSON from an [API].
Future<void> _getHttp() async {
//
  // Get some JSON data from a public API endpoint
  final json = await API.get(
      host: 'boredapi.com',
      path: 'api/activity',
      queryParameters: {'key': '5881028'},
      // headers: headers,
      isHttps: true);

  // print a seperator 70 charcters wide
  Console.seperator(70);
  // prints
  // ======================================================================

  // print the results
  print('Request URL:        ${json['_%url']}');
  print('Response code:      ${json['_%status']}');
  print('Activity:           ${json['activity']}');

// prints:
//
// MY TEST NAME
// ╔═══════════╤═════════════════╤════════════════╗
// ║  term     │       dL        │       cLs      ║
// ╟───────────┼─────────────────┼────────────────╢
// ║  bodrer   │        0.00000  │       1.00000  ║
// ║  board    │        0.26304  │       0.73600  ║
// ║  border   │        0.00000  │       1.00000  ║
// ║  boarder  │        0.22240  │       0.77761  ║
// ║  brother  │        0.22239  │       0.77761  ║
// ║  broad    │        0.26304  │       0.73698  ║
// ║  bored    │        0.26304  │       0.73698  ║
// ╚═══════════╧═════════════════╧════════════════╝

//
}

/// Example of printing JSON data with the [Console] class.
void _printJson(Map<String, Map<String, Object>> jsonCollection) {
  //

  /// print a horizontal double line.
  Console.seperator(80, '=');

  // pass a title and results to the Console.out static method to print results
  Console.out(title: 'MY TEST NAME', results: jsonCollection.values.toList());

  // prints:

// MY TEST NAME
// ╔═══════════╤═════════════════╤════════════════╗
// ║  term     │       dL        │       cLs      ║
// ╟───────────┼─────────────────┼────────────────╢
// ║  bodrer   │        0.00000  │       1.00000  ║
// ║  board    │        0.26304  │       0.73600  ║
// ║  border   │        0.00000  │       1.00000  ║
// ║  boarder  │        0.22240  │       0.77761  ║
// ║  brother  │        0.22239  │       0.77761  ║
// ║  broad    │        0.26304  │       0.73698  ║
// ║  bored    │        0.26304  │       0.73698  ║
// ╚═══════════╧═════════════════╧════════════════╝
}

/// Example of a [JsonDataService] using a Hive Box.
Future<void> _jsonDataServiceExample(
    Map<String, Map<String, Object>> jsonCollection) async {
  //

  // instantiate a JsonDataService
  final service = await JsonDataService.hydrate(
      '${Directory.current.path}\\example', 'sampleStocks');

  // clear the datastore
  await service.dataStore.clear();

  // add all the elements of sampleStocks
  await service.batchUpsert(jsonCollection);

  // read a few records
  final results = (await service.batchRead(['term3', 'term5'])).values.toList();

  // print the records
  Console.out(title: 'JSON DATA SERVICE', results: results);

  // close the service, releasing the resources
  await service.close();

  // prints:
  // ———————————————————————————————————————————————————————————————————————————————————
  // JSON DATA SERVICE
  // ___________________________________________________________________________________
  // │  term                  │             dL             │            cLs            │
  // ———————————————————————————————————————————————————————————————————————————————————
  // │  boarder               │                   0.22240  │                  0.77761  │
  // │  broad                 │                   0.26304  │                  0.73698  │
  // ———————————————————————————————————————————————————————————————————————————————————
}
